presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -287,6 +292,45 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
             return new ExchangeNode(node.getId(), node.getType(), node.getScope(), partitioningScheme, sources, inputs);
         }
 
+        private Symbol canonicalizeExchangeNodeOutput(ExchangeNode node, int symbolIndex, Multimap<Symbol, Symbol> inputToOutputs)
+        {
+            Symbol canonicalOutput = canonicalize(node.getOutputSymbols().get(symbolIndex));
+            Optional<Symbol> outputSymbolForInputSymbols = getExchangeNodeOutputSymbolForInputSymbols(node, symbolIndex, inputToOutputs);
+
+            if (outputSymbolForInputSymbols.isPresent()
+                    && !canonicalOutput.equals(outputSymbolForInputSymbols.get())) {
+                map(canonicalOutput, outputSymbolForInputSymbols.get());
+                return outputSymbolForInputSymbols.get();
+            }
+            else {
+                node.getInputs().stream()
+                        .forEach(input -> inputToOutputs.put(input.get(symbolIndex), canonicalOutput));
+                return canonicalOutput;
+            }
+        }
+
+        private Optional<Symbol> getExchangeNodeOutputSymbolForInputSymbols(ExchangeNode node, int symbolIndex, Multimap<Symbol, Symbol> inputToOutputs)
+        {
+            if (node.getInputs().isEmpty()) {
+                return Optional.empty();
+            }
+
+            Symbol firstSourceCanonicalInput = canonicalize(node.getInputs().get(0).get(symbolIndex));
+            Set<Symbol> outputAliases = new HashSet<>(inputToOutputs.get(firstSourceCanonicalInput));
```

Comment:
why do you put output of `firstSourceCanonicalInput` into `inputToOutputs` for each `symbolIndex`?